contrib/babelfishpg_tsql/src/pltsql_coerce.c

```diff
@@ -888,6 +920,227 @@ run_tsql_best_match_heuristics(int nargs, Oid *input_typeids, FuncCandidateList
 	return new_candidates;
 }
 
+/*
+ * get_immediate_base_type_of_UDT_internal()
+ * This function returns the Immediate base type for UDT.
+ * Returns InvalidOid if given type is not an UDT
+ */
+static Oid
+get_immediate_base_type_of_UDT_internal(Oid typeid)
+{
+	HeapTuple					tuple;
+	bool						isnull;
+	Datum						datum;
+	Datum                       tsql_typename;
+	Oid							base_type;
+	LOCAL_FCINFO(fcinfo, 1);
+
+	/* if common_utility_plugin_ptr is not initialised */
+	if (common_utility_plugin_ptr == NULL)
+		ereport(ERROR,
+				(errcode(ERRCODE_INTERNAL_ERROR),
+					errmsg("Failed to find common utility plugin.")));
+
+	/* if tsql_typename is NULL it implies that inputTypId corresponds to UDT */
+	InitFunctionCallInfoData(*fcinfo, NULL, 0, InvalidOid, NULL, NULL);
+	fcinfo->args[0].value = ObjectIdGetDatum(typeid);
+	fcinfo->args[0].isnull = false;
+	tsql_typename = (*common_utility_plugin_ptr->translate_pg_type_to_tsql) (fcinfo);
+
+	/* if given type is not an UDT then return InvalidOid */
+	if (tsql_typename)
+		return InvalidOid;
+
+	/* Get immediate base type id of given type id */
+	tuple = SearchSysCache1(TYPEOID, ObjectIdGetDatum(typeid));
+	if (!HeapTupleIsValid(tuple))
+		return InvalidOid;
+
+	datum = SysCacheGetAttr(TYPEOID, tuple, Anum_pg_type_typbasetype, &isnull);
+	if (isnull)
+		return InvalidOid;
+
+	base_type = DatumGetObjectId(datum);
+	ReleaseSysCache(tuple);
+
+	return base_type;
+}
+
+PG_FUNCTION_INFO_V1(get_immediate_base_type_of_UDT);
+
+Datum
+get_immediate_base_type_of_UDT(PG_FUNCTION_ARGS)
+{
+	Oid			base_type;
+	
+	base_type = get_immediate_base_type_of_UDT_internal(PG_GETARG_OID(0));
+	if (!OidIsValid(base_type))
+		PG_RETURN_NULL();
+
+	PG_RETURN_OID(base_type);
+}
+
+/*
+ * For a given function details, validate whether it is in special function list
+ * and also validate the input argument data types.
+ */
+bool
+validate_special_function(char *func_nsname, char *func_name, int nargs, Oid *input_typeids)
+{
+	tsql_special_function_t    *special_func;
+	bool                        type_match;
+	Oid                         inputTypId, validTypId, baseTypeId;
+	Oid                         sys_varcharoid;
+
+	/* Sanity checks */
+	if (func_name == NULL || (nargs != 0 && input_typeids == NULL))
+		return false;
+
+	/* 
+	 * Special function handling is only for some specific system functions.
+	 * If func_nsname is NULL, consider it to be a "sys".
+	 */
+	if (func_nsname != NULL &&
+		(strlen(func_nsname) != 3 || strncmp(func_nsname, "sys", 3) != 0))
+		return false;
+
+	/* Get Special function details */
+	special_func = NULL;
+	for (int i = 0; i < TOTAL_TSQL_SPECIAL_FUNCTION_COUNT; i++)
+	{
+		if (strcmp(func_name, tsql_special_function_list[i].funcname) == 0
```

Comment:
Let's update the condition to use ```strncmp``` instead of ```strcmp```

Reply:
Here length of `func_name` is unknown, so there is no specific advantage of using `strncmp` here as strlen will run into same problems faced by `strcmp`.